django_sharding_library/fields.py

```diff
@@ -156,3 +159,30 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
     the shard using a pre_save signal.
     """
     pass
+
+
+class PostgresShardGeneratedIDField(AutoField):
+    """
+    A field that uses a Postgres stored procedure to return an ID generated on the database.
+    """
+    def db_type(self, connection, *args, **kwargs):
+
+        if not hasattr(settings, 'SHARD_EPOCH'):
+            raise ValueError("PostgresShardGeneratedIDField requires a SHARD_EPOCH to be defined in your settings file.")
+
+        if connection.vendor == PostgresDatabaseWrapper.vendor:
+            return "bigint DEFAULT next_sharded_id()"
+        else:
+            return super(PostgresShardGeneratedIDField, self).db_type(connection)
+
+    @staticmethod
+    def migration_receiver(*args, **kwargs):
+        sequence_name = "global_id_sequence"
+        db_alias = kwargs.get('using')
+        if not db_alias:
+            raise EnvironmentError("A pre-migration receiver did not receive a database alias. "
+                                   "Perhaps your app is not registered correctly?")
+        if settings.DATABASES[db_alias]['ENGINE'] == Backends.POSTGRES:
```

Comment:
Yup, you can do it either for all sharded databases or make it a shard
group level setting?

On Tue, May 3, 2016, 8:59 PM Titus Peterson <notifications@github.com>
wrote:

> In django_sharding_library/fields.py
> <https://github.com/JBKahn/django-sharding/pull/26#discussion_r61980198>:
>
> > +        if not hasattr(settings, 'SHARD_EPOCH'):
> > +            raise ValueError("PostgresShardGeneratedIDField requires a SHARD_EPOCH to be defined in your settings file.")
> > +
> > +        if connection.vendor == PostgresDatabaseWrapper.vendor:
> > +            return "bigint DEFAULT next_sharded_id()"
> > +        else:
> > +            return super(PostgresShardGeneratedIDField, self).db_type(connection)
> > +
> > +    @staticmethod
> > +    def migration_receiver(*args, **kwargs):
> > +        sequence_name = "global_id_sequence"
> > +        db_alias = kwargs.get('using')
> > +        if not db_alias:
> > +            raise EnvironmentError("A pre-migration receiver did not receive a database alias. "
> > +                                   "Perhaps your app is not registered correctly?")
> > +        if settings.DATABASES[db_alias]['ENGINE'] == Backends.POSTGRES:
>
> Hmm. The problem is that the library allows arbitrary backends within the
> same shard group. Should I just loop through the entire shard group and
> make sure all the backends are postgres?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/JBKahn/django-sharding/pull/26/files/e534b5ffa93bd79fab7bcb8f90fb315a344f3661#r61980198>
>